### PR TITLE
2669 - Cache-Control no-store returns excessive resources

### PR DESCRIPTION
**Pull request by KevinDougan-SmileCDR**

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/SearchCoordinatorSvcImpl.java [hunk 1]
```diff
@@ -588,8 +588,11 @@ private Integer getQueryCount(Integer theLoadSynchronousUpTo, SearchParameterMap
 	}
 
 	@Nullable
-	private Integer getLoadSynchronousUpToOrNull(CacheControlDirective theCacheControlDirective) {
+	private Integer getLoadSynchronousUpToOrNull(final SearchParameterMap theParams, CacheControlDirective theCacheControlDirective) {
 		final Integer loadSynchronousUpTo;
+		if (theParams != null && !theParams.isEmpty() && theParams.containsKey("_id")) {
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/SearchCoordinatorSvcImpl.java:593]:
```suggestion
		if (theParams != null && theParams.containsKey("_id")) {
```

**Review by fil512 — APPROVED** [review]

**Review by fil512 — DISMISSED** [review]
Per James' feedback, looks like we need to take a different approach to resolve this.

File: hapi-fhir-jpaserver-base/src/test/resources/dstu3/no-store-header/episodeofcare-bundle.json [hunk 2]
```diff
@@ -0,0 +1,4112 @@
+{
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/test/resources/dstu3/no-store-header/episodeofcare-bundle.json:1]: This file looks larger than required to reproduce the bug.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/LegacySearchBuilder.java [hunk 3]
```diff
@@ -804,11 +804,6 @@ public void loadResourcesByPid(Collection<ResourcePersistentId> thePids, Collect
 					pidsToInclude = new HashSet<>(filterResourceIdsByLastUpdated(theEntityManager, theLastUpdated, pidsToInclude));
 				}
 			}
-			for (ResourcePersistentId next : pidsToInclude) {
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/LegacySearchBuilder.java:807]: Was this bug fixed in both the legacy search builder and the new search builder?

**Review by fil512 — APPROVED** [review]

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/LegacySearchBuilder.java (hunk 3 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/LegacySearchBuilder.java:807]: Yup

File: hapi-fhir-jpaserver-base/src/test/resources/dstu3/no-store-header/episodeofcare-bundle.json (hunk 2 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/test/resources/dstu3/no-store-header/episodeofcare-bundle.json:1]: I actually did a pretty meticulous (and tedious :) ) pruning to get it to just what was needed. It's kinda long, but it's a very comprehensive test case this way

**Pull request merged**